docs/standard/native-interop/preserve-sig.md

```diff
@@ -0,0 +1,63 @@
+---
+title: Implicit native signature translations and PreserveSig in .NET native interop
+description: Learn how managed signatures translate to native signatures for interop methods in .NET
+ms.date: 9/21/2023
+---
+
+# Implicit method signature translations in .NET interop
+
+In order to stay programming language agnostic, the Windows COM system and many Windows APIs return a 4 byte integer type called an `HRESULT` to indicate whether an API succeeded or failed, along with some information about the failure. Other values that need to be passed to the caller are "returned" via pointer parameters that act as "out" parameters, and are typically the last parameter in the signature. Languages like C# and Visual Basic traditionally translate a failure code to an exception to match how failures are usually propagated in the language, and expect interop method signatures to not include the `HRESULT`. To translate the method signature to a native signature, the runtime moves the return value of the method to an additional "out" parameter with one more level of indirection (in other words, makes it a pointer to the managed signature's return type), and assumes an `HRESULT` return value. If the managed method returns `void`, no additional parameter is added and the return value becomes an `HRESULT`. For example, see the following two C# COM methods that translate to the same native signature:
+
+```csharp
+int Add(int a, int b);
+
+void Add(int a, int b, out int sum);
+```
+
+```c
+HRESULT Add(int a, int b, /* out */ int* sum);
+```
+
+## PreserveSig in COM
+
+All COM methods in C# are expected to use the translated signature by default. In order to use and export methods without the signature translation and handling of `HRESULT` values, add the <xref:System.Runtime.InteropServices.PreserveSigAttribute> to a COM interface method. When the attribute is applied to a method, no translation is done to the signature, and exceptions will not be thrown for failing `HRESULT` values. This applies to both built-in COM and source-generated COM. For example, see the C# method signature below with a `PreserveSig` attribute and its corresponding native signature.
+
+```csharp
+[PreserveSig]
+int Add(int a, int b, out int sum);
+```
+
+```c
+HRESULT Add(int a, int b, int* sum);
+```
+
+This can be useful if the method may return different `HRESULT` values that are not failures, but must be handled differently. For example, some methods may return the value `S_FALSE` when a method does not fail, but only returns partial results, and `S_OK` when it returns all results.
```

Comment:
```suggestion
This can be useful if the method might return different `HRESULT` values that aren't failures, but must be handled differently. For example, some methods might return the value `S_FALSE` when a method doesn't fail but only returns partial results, and `S_OK` when it returns all results.
```